## CIND830 - Python Programming for Data Science  
### Assignment 2 (10% of the final grade) 
### Due on March 07, 2022 11:59 PM 

*****
This is a Jupyter Notebook document that extends a simple formatting syntax for authoring HTML and PDF. Review [this](https://jupyter-notebook.readthedocs.io/en/stable/notebook.html) website for more details on using Jupyter Notebooks.

Use the JupyterHub server on the Google Cloud Platform, provided by your designated instructor, for this assignment. Ensure using **Python 3.7** release or higher then complete the assignment by inserting your Python code wherever seeing the string `#INSERT YOUR ANSWER HERE.`

When you click the `File` button, from the top navigation bar, then select `Export Notebook to HTML`, an HTML document will be generated that includes both the assignment content and the output of any embedded Python code chunks.

Use [these](https://www.ryerson.ca/courses/students/tutorials/assignments/) guidelines to submit **both** the IPYNB and the exported file (HTML). Failing to submit both files will be subject to mark deduction.

Please be advised that you cannot get more than 100% in this assignment, and the **BONUS** question (if there is any) will only be graded if all other questions have been submitted.

*****


### Question 1  **[25 pts]**:

**a)**  **[10 pts]**  Write a Python program that converts a list of email addresses to a dictionary, where the keys are the usernames, and the values are the respective domains.

For example:  If the user enters `harry@abc.com,larry@abc.ca,sally@abc.org`, then the output should be `{'harry': 'abc.com', 'larry': 'abc.ca', 'sally': 'abc.org'}`


In [1]:
email_dict = {}
  
while True:
    email = str(input("Enter email address or skip to show dictionary: "))
    if '@' in email:
        email_list = email.split('@')
        username = email_list[0]
        domain = email_list[1]
        email_dict[username] = domain
    elif email == "":
        print("Email dictionary:\n",email_dict)
        break
    else:
        print("Error: Invalid email address.")


Enter email address or skip to show dictionary: ana@hello.ca
Enter email address or skip to show dictionary: banana@hi.ca
Enter email address or skip to show dictionary: 
Email dictionary
 {'ana': 'hello.ca', 'banana': 'hi.ca'}


**b)**  **[10 pts]** Write a Python program to accept the length of a dictionary, then create key-value pairs where the key is a random integer in the range 1 to 100 and the value is the square of the integer.

For example: If the user enters `5`, then the output should be `{68: 4624, 28: 784, 99: 9801, 15: 225, 6: 36}`

In [2]:
import random
dict = {}
length = int(input("Enter desired number of random integers: "))

for i in range(0,length):
    key = random.randint(1,100)
    dict[key] = key ** 2

print(dict)

Enter desired number of random integers: 5
{98: 9604, 72: 5184, 12: 144, 21: 441}


**c)**  **[5 pts]** List at least three diferences between a list and a dictionary.  Illustrate your work with simple examples.

In [3]:
list1 = ['P','Y','T','H','O','N']
dict1 = {}
dict1['zero'] = 'P'
dict1[1] = 'Y'
dict1['two'] = 'T'
dict1[(3,0)] = 'H'
dict1['four'] = 'O'
dict1[5] = 'N'

#1) A list is a data structure wherein each element within it has an index integer. Meanwhile, a dictionary in Python uses 'keys' rather than integers for indexing.
#   Keys could be an integer, but it could also be a string or a tuple. Any immutable data type can be a key.

print(list1[3]) #The index must strictly be an integer.
print(dict1[(3,0)]) #The index or key in this case is a tuple.

#2) To add a new element in a list, methods such as .append(), .extend(), and .insert() may be used.
#   For dictionaries, adding key-value pairs does not need a method.

list1.append('!')
dict1[(6,0)] = '!'

#3) It is straightforward and easy to see all the elements of a list. Sorting them requires very little effort as well.
#   There is no direct way to see all elements of a dictionary. Either a for loop is used to print each key-value pair, or the dictionary must be converted to a list.

print(list1)

for x in dict1:
    print('Key: ', x, '\tValue: ', dict1[x])
# or
list(dict1.items())

H
H
['P', 'Y', 'T', 'H', 'O', 'N', '!']
Key:  zero 	Value:  P
Key:  1 	Value:  Y
Key:  two 	Value:  T
Key:  (3, 0) 	Value:  H
Key:  four 	Value:  O
Key:  5 	Value:  N
Key:  (6, 0) 	Value:  !


[('zero', 'P'),
 (1, 'Y'),
 ('two', 'T'),
 ((3, 0), 'H'),
 ('four', 'O'),
 (5, 'N'),
 ((6, 0), '!')]

*****

### Question 2  **[25 pts]**:

The data file used in this question is downloaded from (https://www.kaggle.com/aungpyaeap/supermarket-sales) and converted to a CSV file.  The [`supermarket_sales.csv`](https://app.box.com/s/f4hcfkferizntbev3ou8hso6nyfccf74) is a delimited text file containing the historical record of sales data in three supermarkets. The file contains the following columns separated by the comma character `,`
* Invoice id: Computer-generated sales slip invoice identification number
* Branch: Branch of supercenter (3 branches are available identified by A, B and C)
* City: Location of supercenters
* Customer type: Type of customers, recorded by Members for customers using member card and Normal for without member card
* Gender: Gender type of customer
* Product line: General item categorization groups - Electronic accessories, Fashion accessories, Food and beverages, Health and beauty, Home and Lifestyle, Sports and travel
* Unit price: The price of each product in $
* Quantity: Number of products purchased by the customer
* Tax: 5% tax fee for customer buying
* Total: Total price including tax
* Date: Date of purchase (Record available from January 2019 to March 2019)
* Time: Purchase time (10am to 9pm)
* Payment: Payment used by customer for purchase (3 methods are available – Cash, Credit card and Ewallet)
* COGS: Cost of goods sold
* Gross margin percentage: Gross margin percentage
* Gross income: Gross income
* Rating: Customer satisfaction rating on their overall shopping experience (On a scale of 1 to 10)

Note: Only Python [built-in functions](https://docs.python.org/3.7/library/functions.html) or [standard modules](https://docs.python.org/3.7/py-modindex.html) are allowed to be used in this question. Do not use a third-party library for any part of the code. Importing any third-party library will result in marks deduction.

**a)**  **[10 pts]**  Write a Python function that reads the Supermarket Sales data into a 2 dimensional structure.

Note: The first line of the CSV contains the column headings, therefore it should not be stored in the 2 dimensional data list. You can skip the first line of the data file

In [6]:
import csv

def csvreader(x):
    header = next(x)
    saleslist = list(x)
    return saleslist

with open('supermarket_sales.csv', 'r') as csvfile:
    salesreader = csv.reader(csvfile,delimiter=',')
    sales = csvreader(salesreader)
    print("Data:\n", sales)

Data:
 [['750-67-8428', 'A', 'Yangon', 'Member', 'Female', 'Health and beauty', '74.69', '7', '26.1415', '548.9715', '1/5/2019', '13:08', 'Ewallet', '522.83', '4.761904762', '26.1415', '9.1'], ['226-31-3081', 'C', 'Naypyitaw', 'Normal', 'Female', 'Electronic accessories', '15.28', '5', '3.82', '80.22', '3/8/2019', '10:29', 'Cash', '76.4', '4.761904762', '3.82', '9.6'], ['631-41-3108', 'A', 'Yangon', 'Normal', 'Male', 'Home and lifestyle', '46.33', '7', '16.2155', '340.5255', '3/3/2019', '13:23', 'Credit card', '324.31', '4.761904762', '16.2155', '7.4'], ['123-19-1176', 'A', 'Yangon', 'Member', 'Male', 'Health and beauty', '58.22', '8', '23.288', '489.048', '1/27/2019', '20:33', 'Ewallet', '465.76', '4.761904762', '23.288', '8.4'], ['373-73-7910', 'A', 'Yangon', 'Normal', 'Male', 'Sports and travel', '86.31', '7', '30.2085', '634.3785', '2/8/2019', '10:37', 'Ewallet', '604.17', '4.761904762', '30.2085', '5.3'], ['699-14-3026', 'C', 'Naypyitaw', 'Normal', 'Male', 'Electronic accessories'

**b)**  **[5 pts]**  Write a function to receive the Supermarket Sales list (that you created in the previous question) and return the total sales (the sum of column `Total`) of the branches as a dictionary

The returned information should look like (the numbers given below may not be correct):  
{"A": 123456.123, "B": 234567.567, "C": 345678.9987}

In [7]:
def salessum(x):
    sum_dict = {}
    sum_A = 0
    sum_B = 0
    sum_C = 0
    for i in x:
        if i[1] == 'A':
            sum_A += float(i[9])
        elif i[1] == 'B':
            sum_B += float(i[9])
        elif i[1] == 'C':
            sum_C += float(i[9])
    sum_dict['A'] = sum_A
    sum_dict['B'] = sum_B
    sum_dict['C'] = sum_C
    return sum_dict

salessum(sales)

{'A': 106200.3705000001, 'B': 106197.67199999996, 'C': 110568.70649999994}

**c)  [5 pts]** Write a function to receive the Supermarket Sales list and return an object (maybe a list or a dictionary) containing the total number of sales per Branch and per Customer Type.

The returned information can be either one of the below choices or another (The numbers given below may not be correct):)
* a dictionary like {"A": {"Member": 230, "Normal": 351}, "B": {"Member": 123, "Normal": 117},  "C": {"Member": 335, "Normal": 18}}
* a list like [ [230, 351], [123, 117], [335, 18]]

You need to consider the best structure to return since you will be displaying this information in a human-readable format in the following question. 

In [8]:
def salesnum(x):
    allbranch = {}
    branch_A = {}
    branch_B = {}
    branch_C = {}
    memcount_A = 0
    normcount_A = 0
    memcount_B = 0
    normcount_B = 0
    memcount_C = 0
    normcount_C = 0

    for i in x:
        if i[1] == 'A' and i[3] == 'Member':
            memcount_A += 1
        elif i[1] == 'A' and i[3] == 'Normal':
            normcount_A += 1
        elif i[1] == 'B' and i[3] == 'Member':
            memcount_B += 1
        elif i[1] == 'B' and i[3] == 'Normal':
            normcount_B += 1
        elif i[1] == 'C' and i[3] == 'Member':
            memcount_C += 1
        elif i[1] == 'C' and i[3] == 'Normal':
            normcount_C += 1
    
    branch_A['Member'] = memcount_A
    branch_A['Normal'] = normcount_A
    branch_B['Member'] = memcount_B
    branch_B['Normal'] = normcount_B
    branch_C['Member'] = memcount_C
    branch_C['Normal'] = normcount_C

    allbranch['A'] = branch_A
    allbranch['B'] = branch_B
    allbranch['C'] = branch_C

    return allbranch

print(salesnum(sales))

{'A': {'Member': 167, 'Normal': 173}, 'B': {'Member': 165, 'Normal': 167}, 'C': {'Member': 169, 'Normal': 159}}


**d) [5 pts]** Write a Python program to do the following:
* Call the function you wrote in **a)** to read the data file
* Call the function you wrote in **b)** to calculate the total sales information and display the information on the screen
* Call the function you wrote in **c)** to calculate the total number of sales per Branch and per Customer Type and display the information on the screen

Note: While displaying the information on the screen, the outputs should be human-readable and understandable (e.g. column headings, row names)

In [100]:
import csv

with open('supermarket_sales.csv', 'r') as csvfile:
    salesreader = csv.reader(csvfile,delimiter=',')
    headreader = csv.DictReader(csvfile)
    headers = headreader.fieldnames
    sales = csvreader(salesreader)

headers.insert(0, '#')

for j in headers:
    if j == headers[0]:
        print(j, end='\t')
        continue
    print(j, end='\t\t')

print()

for k in sales:
    print(sales.index(k)+1, end='\t')
    for l in k:
        print(l, end='\t\t')
    print("\n----------")

print()

print("Total Sales per Branch")
display_salessum = salessum(sales)
for m in display_salessum:
    print(m,"\t","$",round(display_salessum[m],2))

print()

print("Total Sale Transactions per Branch")
display_salesnum = salesnum(sales)
for n in display_salesnum:
    for o in n:
        print(n, display_salesnum[o])

#	Invoice ID		Branch		City		Customer type		Gender		Product line		Unit price		Quantity		Tax 5%		Total		Date		Time		Payment		cogs		gross margin percentage		gross income		Rating		
1	226-31-3081		C		Naypyitaw		Normal		Female		Electronic accessories		15.28		5		3.82		80.22		3/8/2019		10:29		Cash		76.4		4.761904762		3.82		9.6		
----------
2	631-41-3108		A		Yangon		Normal		Male		Home and lifestyle		46.33		7		16.2155		340.5255		3/3/2019		13:23		Credit card		324.31		4.761904762		16.2155		7.4		
----------
3	123-19-1176		A		Yangon		Member		Male		Health and beauty		58.22		8		23.288		489.048		1/27/2019		20:33		Ewallet		465.76		4.761904762		23.288		8.4		
----------
4	373-73-7910		A		Yangon		Normal		Male		Sports and travel		86.31		7		30.2085		634.3785		2/8/2019		10:37		Ewallet		604.17		4.761904762		30.2085		5.3		
----------
5	699-14-3026		C		Naypyitaw		Normal		Male		Electronic accessories		85.39		7		29.8865		627.6165		3/25/2019		18:30		Ewallet		597.73		4.761904762		29.8865		4.1		
----------
6	355-53-5

*****

### Question 3  **[25 pts]**: 
Write a program detailing the number of passengers per vehicle.  The program includes three classes: `Vehicle`, `Car`, `Bus` where `Vehicle` is the parent class of `Car` and `Bus`. 

**a)**  **[10 pts]**  Each of the `Car` and `Bus` classes should have a constructor method `__init()__` that generates and returns a random number of passengers based on the type of the vehicle; If it is a `Bus`, then the number of passengers should be within a range of 5 and 20.  If the vehicle is a `Car`, the range should be between 1 to 4 passengers, inclusively.

**b)**  **[10 pts]**  The program's `main()` function would display the following output.  Notably, since the numbers of passengers are randomly selected, the program will keep changing them every time it runs.

```
--------------------------------
Bus Passengers:   64
Car Passengers:   13
Total Passengers: 77
--------------------------------
Bus#:  1  Number of Passengers :18
Bus#:  2  Number of Passengers :20
Bus#:  3  Number of Passengers : 6
Bus#:  4  Number of Passengers : 9
Bus#:  5  Number of Passengers :11
Car#:  6  Number of Passengers : 1
Car#:  7  Number of Passengers : 2
Car#:  8  Number of Passengers : 4
Car#:  9  Number of Passengers : 2
Car#: 10  Number of Passengers : 4
--------------------------------
```

**c)**  **[5 pts]** Restructure the program by re-defining the `Bus` class to be the superclass of `Car`, and display the same output.

In [92]:
# Code for subquestions (a) and (b)
import random

class Vehicle:
    def __init__(self):
        self.passenger = 0

class Bus(Vehicle):
    def __init__(self):
        self.passenger = random.randint(1,20)
    def getpassenger(self):
        return self.passenger

class Car(Vehicle):
    def __init__(self):
        self.passenger = random.randint(1,4)
    def getpassenger(self):
        return self.passenger
    
def main(): 

    busiterations = []
    cariterations = []
    bustotal = 0
    cartotal = 0
    
    for j in range(1,6):
        bus_instance = Bus()
        busiterations.append([j,bus_instance.getpassenger()])
        bustotal += bus_instance.getpassenger()
    for k in range(6,11):
        car_instance = Car()
        cariterations.append([k,car_instance.getpassenger()])
        cartotal += car_instance.getpassenger()
    
    # prints results
    for i in range(34):
        print("-",end='')
    print()
    print("Bus Passengers:", "%5s" % bustotal)
    print("Car Passengers:", "%5s" % cartotal)
    print("Total Passengers:", "%3s" % (bustotal + cartotal))
    for i in range(34):
        print("-",end='')
    print()
    for l in busiterations:
        print("Bus#:","%2s" % l[0]," Number of Passengers:","%2s" % l[1])
    for m in cariterations:
        print("Car#:","%2s" % m[0]," Number of Passengers:","%2s" % m[1])
    for i in range(34):
        print("-",end='')
    print()
    
if __name__ == '__main__':
    main()

----------------------------------
Bus Passengers:    59
Car Passengers:    12
Total Passengers:  71
----------------------------------
Bus#:  1  Number of Passengers: 13
Bus#:  2  Number of Passengers: 10
Bus#:  3  Number of Passengers: 17
Bus#:  4  Number of Passengers: 14
Bus#:  5  Number of Passengers:  5
Car#:  6  Number of Passengers:  1
Car#:  7  Number of Passengers:  3
Car#:  8  Number of Passengers:  4
Car#:  9  Number of Passengers:  3
Car#: 10  Number of Passengers:  1
----------------------------------


In [98]:
# code for subquestion (c)

import random

class Vehicle:
    def __init__(self):
        self.passenger = 0

class Bus(Vehicle):
    def __init__(self):
        self.passenger = random.randint(1,20)
    def getpassenger(self):
        return self.passenger

class Car(Bus):
    def __init__(self):
        super().__init__()
        self.passenger = random.randint(1,4)

def main(): 

    busiterations = []
    cariterations = []
    bustotal = 0
    cartotal = 0
    
    
    for j in range(1,6):
        bus_instance = Bus()
        busiterations.append([j,bus_instance.getpassenger()])
        bustotal += bus_instance.getpassenger()
    for k in range(6,11):
        car_instance = Car()
        cariterations.append([k,car_instance.getpassenger()])
        cartotal += car_instance.getpassenger()
    
    # prints results
    for i in range(34):
        print("-",end='')
    print()
    print("Bus Passengers:", "%5s" % bustotal)
    print("Car Passengers:", "%5s" % cartotal)
    print("Total Passengers:", "%3s" % (bustotal + cartotal))
    for i in range(34):
        print("-",end='')
    print()
    for l in busiterations:
        print("Bus#:","%2s" % l[0]," Number of Passengers:","%2s" % l[1])
    for m in cariterations:
        print("Car#:","%2s" % m[0]," Number of Passengers:","%2s" % m[1])
    for i in range(34):
        print("-",end='')
    print()
    
if __name__ == '__main__':
    main()

----------------------------------
Bus Passengers:    69
Car Passengers:    10
Total Passengers:  79
----------------------------------
Bus#:  1  Number of Passengers: 11
Bus#:  2  Number of Passengers: 17
Bus#:  3  Number of Passengers: 15
Bus#:  4  Number of Passengers: 11
Bus#:  5  Number of Passengers: 15
Car#:  6  Number of Passengers:  2
Car#:  7  Number of Passengers:  2
Car#:  8  Number of Passengers:  2
Car#:  9  Number of Passengers:  1
Car#: 10  Number of Passengers:  3
----------------------------------


*****

### Question 4  **[25 pts]**:

**a)**  **[5 pts]** Define and implement a function that accepts a tuple of lists and returns the sum of only the unique numbers in that tuple.

For example, if the user enters the following tuple: `
([1, 5], [7, 8, 3, 2], [3, 3, 2])`
then the output should be `26`.

In [4]:
def unique(x):
    x_list = []
    x_unique = []
    sum = 0

    for i in list(x): #converts the input tuple of lists into a simple list
        for j in i:
            x_list.append(j)
    for k in x_list: #creates a list of unique values
        if k not in x_unique:
            x_unique.append(k)
    for l in x_unique: #gets the sum of unique values
        sum = sum + int(l)
    
    return sum

entry_tolist = []
list_totuple = []
tuple_fromlist = ()

while True:
    entry = input("Input a list of numbers, separated by a comma or skip to process:\n")
    if entry == "":
        tuple_fromlist = tuple(list_totuple)
        print("Tuple generated:\t",tuple_fromlist)
        print("Sum of unique values:\t",unique(tuple_fromlist))
        break
    else:
        entry_tolist = entry.split(",")
        list_totuple.append(entry_tolist)


Input a list of numbers, separated by a comma or skip to process:
1,3
Input a list of numbers, separated by a comma or skip to process:
2,3
Input a list of numbers, separated by a comma or skip to process:
4,5
Input a list of numbers, separated by a comma or skip to process:

Tuple generated:	 (['1', '3'], ['2', '3'], ['4', '5'])
Sum of unique values:	 15


**b)**  **[5 pts]** Define and implement a function that converts the a list of tuples to a dictionary where the first value of each tuple is the key, and the second value is the dictionary value. <br>

For example, if the user enters the following list: 
`
[('Person1', 80), ('Person2', ), ('Person3', 20), ('Person4', 20, 60)]
`

Then the output should be: `{'Person1': 80, 'Person2': None, 'Person3': 20, 'Person4': 20}`


In [ ]:
def create_dict(x):
    newdict = {}
    for i in x:
        j = list(i)
        if len(j) == 1: #adds the value None if list only has one element
            j.append(None)
        newdict[j[0]] = j[1]
    return newdict

tuples_list = []

while True:
    entry = str(input("Enter a two-element tuple separated by comma, or skip to process:\n"))
    if entry == "":
        print("Tuple entered:\t",tuples_list)
        print("Dictionary created:\t",create_dict(tuples_list))
        break
    else:
        entry_tolist = entry.split(",")
        entry_totuple = tuple(entry_tolist)
        tuples_list.append(entry_totuple)

Enter a two-element tuple separated by comma, or skip to process:
1,ana
Enter a two-element tuple separated by comma, or skip to process:
2,pat
Enter a two-element tuple separated by comma, or skip to process:
3,gaw
Enter a two-element tuple separated by comma, or skip to process:
4
Enter a two-element tuple separated by comma, or skip to process:
5,abe
Enter a two-element tuple separated by comma, or skip to process:

Tuple entered:	 [('1', 'ana'), ('2', 'pat'), ('3', 'gaw'), ('4',), ('5', 'abe')]
Dictionary created:	 {'1': 'ana', '2': 'pat', '3': 'gaw', '4': None, '5': 'abe'}


**c)**  **[15 pts]** The following script extracts the text from the `https://norvig.com/big.txt` file, stores them in a string object, then assigns this string to a variable called `data` using the [requests](https://pypi.org/project/requests/) third-party library.
```
#!pip install requests
import requests
response = requests.get('https://norvig.com/big.txt')
data = response.text
```
Run the script above, uncomment the first line if you need to install the library on your machine, then define and implement a function that returns the followings:

* The total number of unique words and unique characters
* The top ten most frequent words and the number of times these words appear in the text.

In [101]:
!pip install requests
import requests
response = requests.get('https://norvig.com/big.txt')
data = response.text
data = data.lower()
word_list = data.split(" ")

from collections import Counter
# Please note, I tried brute force (see bottom) but runtime went over 30 minutes, so I went for the more time-efficient method.
# We were explicity told not to use third-party libraries for Q#2, but nothing was said for Q#4. :)

def countunique(x, y):
    uchar = len(set(x))
    uword = len(set(y))
    return(uword,uchar)

def topunique(z):
    word_count = Counter(z)
    top10_list = word_count.most_common(11)
    return(top10_list)

print("")
print("Total unique words: ",countunique(data,word_list)[0])
print("Total unique characters: ",countunique(data,word_list)[1])

print("")
print("----10 Most Frequent Words----")
for i in topunique(word_list):
    if i[0] == '':
        continue
    print("Word: ",i[0],"\tCount: ",i[1])

print("")

#Below was my original script for brute force, if needed.
'''for uword in unique_list:
  count = 0
  for word in word_list:
    if word == uword:
      count += 1
  countuniques[uword] = count

sort_countuniques = sorted(countuniques.items(), key=lambda x: x[1], reverse=True)

for j in range(10):
  print(sort_countuniques[j])'''


Total unique words:  139448
Total unique characters:  67

----10 Most Frequent Words----
Word:  the 	Count:  65168
Word:  of 	Count:  35009
Word:  and 	Count:  31776
Word:  to 	Count:  25176
Word:  in 	Count:  18741
Word:  a 	Count:  18118
Word:  that 	Count:  10035
Word:  he 	Count:  9998
Word:  was 	Count:  9780
Word:  his 	Count:  8528



'for uword in unique_list:\n  count = 0\n  for word in word_list:\n    if word == uword:\n      count += 1\n  countuniques[uword] = count\n\nsort_countuniques = sorted(countuniques.items(), key=lambda x: x[1], reverse=True)\n\nfor j in range(10):\n  print(sort_countuniques[j])'

*****

#### This is the end of assignment 2